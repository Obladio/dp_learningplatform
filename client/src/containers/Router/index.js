import React, {Component} from "react"
import {connect} from "react-redux"
import {push} from "redux-little-router"

import {routes} from "../../config"

import "./styles.css"

class Router extends Component {
  flattenedRoutes = {}
  flattenRoutes(routes, parentRoute = "") {
    Object.entries(routes).forEach(([route, item]) => {
      if (typeof item !== "object") {
        if (typeof this.flattenedRoutes[parentRoute + route] === "undefined") {

          const newRoute = {},
            routeIdentifier = (parentRoute === "/" && process.env.PUBLIC_URL !== "" ? process.env.PUBLIC_URL : process.env.PUBLIC_URL + parentRoute)

          newRoute[routeIdentifier] = routes

          this.flattenedRoutes = { ...this.flattenedRoutes, ...newRoute }
        }
      }
      else this.flattenRoutes(item, parentRoute + route)
    })
  }

  // getRouteContainer() {
  //   return Object.entries(routes).map(([route, item]) => {
  //     if (!(item.container.prototype instanceof React.Component || item.container.prototype instanceof React.PureComponent)) {
  //       throw new Error(
  //         `Container for route "${route}" must be a React component or pure component`
  //       )
  //     }

  //     if(route === this.props.router.route) {
  //       // show route container if route is not private or token is available
  //       if(!item.private || (item.private && this.props.token !== null)) {
  //         return <item.container key={item.key} />
  //       }
  //     }

  //     return null
  //   })
  // }

  render() {
    this.flattenRoutes(routes)
    return Object.entries(this.flattenedRoutes).map(([route, item]) => {
      if (!(item.container.prototype instanceof React.Component || item.container.prototype instanceof React.PureComponent)) {
      // if (typeof item.container !== "undefined" && !(item.container.prototype instanceof React.Component)) {
        throw new Error(
          `Container for route "${route}" must be a React component or pure component`
        )
      }

      if (route === this.props.router.route) {
        // show route container if route is not private or token is available
        if (!item.private || (item.private && this.props.token !== null)) {
          return <item.container key={item.key} />
        }
      }

      return null
    })
  }
}

const mapStateToProps = state => {
  return {
    router: state.router,
    token: state.login.token
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    push: (location) => dispatch(push(location))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Router)
