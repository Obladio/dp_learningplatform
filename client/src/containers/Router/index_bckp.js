import React, {Component} from "react"
import {connect} from "react-redux"
import {push} from "redux-little-router"

import {routes} from "../../config"

import styles from "./styles.css"

class Router extends Component {

  getRouteContainer() {
    return Object.entries(routes).map(([route, item]) => {
      if (!(item.container.prototype instanceof React.Component || item.container.prototype instanceof React.PureComponent)) {
        throw new Error(
          `Container for route "${route}" must be a React component or pure component`
        )
      }

      if(route === this.props.router.route) {
        // show route container if route is not private or token is available
        if(!item.private || (item.private && this.props.token !== null)) {
          return <item.container key={item.key} />
        }
      }

      return null
    })
  }

  render() {
    return (
      <div className={styles.Container}>
        {this.getRouteContainer()}
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    router: state.router,
    token: state.login.token
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    push: (location) => dispatch(push(location))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Router)
