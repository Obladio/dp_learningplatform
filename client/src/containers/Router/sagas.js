import {select, call} from "redux-saga/effects"
import {push} from "redux-little-router"

import {routes} from "../../config"
import { i18n } from "../../utils"

let flattenedRoutes = {}

const flattenRoutes = (routes, parentRoute = "") => {
  Object.entries(routes).forEach(([route, item]) => {
    if (typeof item !== "object") {
      if (typeof flattenedRoutes[parentRoute + route] === "undefined") {

        const newRoute = {},
          routeIdentifier = (parentRoute === "/" && process.env.PUBLIC_URL !== "" ? process.env.PUBLIC_URL : process.env.PUBLIC_URL + parentRoute)

        newRoute[routeIdentifier] = routes

        flattenedRoutes = { ...flattenedRoutes, ...newRoute }
      }
    }
    else flattenRoutes(item, parentRoute + route)
  })
}

export default function* router(dispatch, action) {
  const {payload} = action
  const {route} = payload
  const state = yield select()

  // change language based on route param
  if (typeof state.router.params.lang !== "undefined") {
    i18n.changeLanguage(state.router.params.lang)
  }

  flattenRoutes(routes)

  for (let routePath in flattenedRoutes) {
    if (routePath !== route) {
      continue
    }

    let routeParams = flattenedRoutes[routePath]

    // check if route is private
    // if so, redirect to login
    if(routeParams.private && state.login.token === null) {
      dispatch(push(`/de/login?referer=${encodeURI(state.router.pathname)}`))

    // otherwise call route saga
    } else {
      if (typeof routeParams.sagas !== "undefined") {
        // scroll window to top on router change
        window.scrollTo(0, 0)

        yield call(routeParams.sagas, dispatch, action)
      }
    }
  }
}
