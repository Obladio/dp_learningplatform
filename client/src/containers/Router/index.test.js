import React from "react"
import ReactDOM from "react-dom"
import renderer from "react-test-renderer"
import {Provider} from "react-redux"

import createStore from "../../redux"
import Router from "./"

const boilerplateStore = createStore()
const store = {
  ...boilerplateStore,
  getState: () => {
    return {
      ...boilerplateStore.getState(),
      router: {
        route: "/"
      }
    }
  }
}

it("renders without crashing", () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router />
    </Provider>
  )

  expect(tree).toMatchSnapshot()
})
