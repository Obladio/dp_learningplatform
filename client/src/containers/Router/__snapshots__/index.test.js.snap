// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className={undefined}
>
  <div
    className="container-default"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <h1>
          Categories
        </h1>
      </div>
    </div>
  </div>
</div>
`;
